lib/AST/Stmt.cpp

```diff
@@ -428,6 +428,9 @@ void ForEachStmt::setPattern(Pattern *p) {
 }
 
 Expr *ForEachStmt::getTypeCheckedSequence() const {
+  if (auto *expansion = dyn_cast<PackExpansionExpr>(getParsedSequence()))
+    return expansion;
```

Comment:
I think we need to do more checking here or add a bit to the statement to make sure that we are only returning non-null pack expansion expressions iff type-checking was successful. For regular for-in statements type-checker sets `iteratorVar`, for pack iteration I think we just need to tag the expression itself (no harm if that applies for both packs and regular sequences too!).

Reply:
I think that we might need to rethink the purpose of this method with this feature in mind. The bug where pack iteration was not working in closures had to partially deal with the fact that whoever calls it assumes that the type-checking was successful **and** we are dealing with a sequence if the method is non-null: https://github.com/apple/swift/pull/67594/commits/90ca95c387f613f0b3a20fa93d22ea6ef1868411#diff-f7b20ead68204a38f1ecf3cd2202f98fbcbfc193e117458b8ce6e612cb8855c7R1893. When we are dealing with a pack expansion, the assumed "sequence" will be non-null (but containing an expansion expr) and we will go into the sequence code path, which is undesired. So the only way to use this method "right" is to check if casting the result into an expansion expression is non-null to make sure that the sequence is null... Which is not obvious at all and we need to fix this. I think I'll put a TODO on it for now